### PR TITLE
638: Per region app policy

### PR DESCRIPTION
**Pull request by f1sh1918**
Find the acceptance criteria and technical tasks to check if everything works as expected [here](https://github.com/digitalfabrik/entitlementcard/issues/638)

Testing requires a Drop table of `regions` table and reimport (at least in couldn't find better solution)

### CODE REVIEW

File: backend/src/main/kotlin/app/ehrenamtskarte/backend/regions/webservice/schema/RegionsQueryService.kt [hunk 1]
```diff
@@ -11,15 +11,20 @@ class RegionsQueryService {
...
             if (it == null) null
-            else Region(it.id.value, it.prefix, it.name, it.regionIdentifier)
+            else Region(it.id.value, it.prefix, it.name, it.regionIdentifier, it.dataPrivacyPolicy)
         }
     }
+
+    @GraphQLDescription("Returns region data for specific region.")
+    fun regionByRegionId(regionId: Int): Region = transaction {
+        Region(RegionsRepository.findRegionById(regionId).id.value, RegionsRepository.findRegionById(regionId).prefix, RegionsRepository.findRegionById(regionId).name, RegionsRepository.findRegionById(regionId).regionIdentifier, RegionsRepository.findRegionById(regionId).dataPrivacyPolicy)
+    }
```
Comment (michael-markl) [inline backend/src/main/kotlin/app/ehrenamtskarte/backend/regions/webservice/schema/RegionsQueryService.kt:29]: Something like this should also work:
```suggestion
    fun regionByRegionId(regionId: Int): Region = transaction {
        val regionEntity = RegionsRepository.findRegionById(regionId)
        
        Region(regionEntity.id.value, regionEntity.prefix, regionEntity.name, regionEntity.regionIdentifier, regionEntity.dataPrivacyPolicy)
    }
```

File: administration/src/constants/dataPrivacyBase.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
+export const dataPrivacyBaseText =
+  'Diese Datenschutzerklärung bezieht sich auf die Verarbeitung personenbezogener Daten im Rahmen der digitalen bayerischen Ehrenamtskarte einschließlich verbundener Dienste.\n\n' +
+  'Die digitale bayerischen Ehrenamtskarte wird im Auftrag des Bayerischen Staatsministeriums für Familie, Arbeit und Soziales, Winzererstraße 9, 80797 München durch die Tür an Tür – Digitalfabrik gGmbH, Wertachstr. 29, 86153 Augsburg betrieben.\n\n' +
```
Comment (michael-markl) [inline administration/src/constants/dataPrivacyBase.ts:3]:
```suggestion
  'Die digitale bayerische Ehrenamtskarte wird im Auftrag des Bayerischen Staatsministeriums für Familie, Arbeit und Soziales, Winzererstraße 9, 80797 München durch die Tür an Tür – Digitalfabrik gGmbH, Wertachstr. 29, 86153 Augsburg betrieben.\n\n' +
```

**Review by michael-markl — CHANGES_REQUESTED** [review]

File: administration/src/application/components/forms/ApplicationForm.tsx [hunk 3]
```diff
@@ -1,10 +1,19 @@
...
-import { FormControl, FormControlLabel, FormLabel, Radio, RadioGroup } from '@mui/material'
+import { Button, Checkbox, FormControl, FormControlLabel, FormGroup, FormLabel, Radio, RadioGroup } from '@mui/material'
 import { Form } from '../../FormType'
 import StandardEntitlementForm, { StandardEntitlementFormState } from './StandardEntitlementForm'
 import PersonalDataForm, { PersonalDataFormState } from './PersonalDataForm'
+import BasicDialog from '../BasicDialog'
+import { dataPrivacyBaseHeadline, dataPrivacyBaseText } from '../../../constants/dataPrivacyBase'
+import styled from 'styled-components'
+
+const PrivacyPolicyButton = styled(Button)`
```
Comment (maxammann) [inline administration/src/application/components/forms/ApplicationForm.tsx:12]: @michael-markl are we actually using styled-components? I can not remember using it.

File: administration/src/application/components/forms/ApplicationForm.tsx [hunk 4]
```diff
@@ -1,10 +1,19 @@
...
+  text-transform: capitalize !important;
```
Comment (maxammann) [inline administration/src/application/components/forms/ApplicationForm.tsx:13]: Is important! requried here?

File: backend/src/main/kotlin/app/ehrenamtskarte/backend/regions/database/repos/RegionsRepository.kt [hunk 5]
```diff
@@ -25,4 +25,12 @@ object RegionsRepository {
 
     fun findByIds(ids: List<Int>) =
         RegionEntity.find { Regions.id inList ids }.sortByKeys({ it.id.value }, ids)
+
+    fun findRegionById(regionId: Int): RegionEntity {
+        return RegionEntity[regionId]
+    }
+
+    fun updateDataPolicy(region: RegionEntity, dataPrivacyText: String) {
```
Comment (maxammann) [inline backend/src/main/kotlin/app/ehrenamtskarte/backend/regions/database/repos/RegionsRepository.kt:33]: Check length of dataPrivacyText here to avoid errors in the DB.

File: backend/src/main/kotlin/app/ehrenamtskarte/backend/regions/webservice/schema/RegionsMutationService.kt [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+import app.ehrenamtskarte.backend.regions.database.repos.RegionsRepository
+import com.expediagroup.graphql.generator.annotations.GraphQLDescription
+import org.jetbrains.exposed.sql.transactions.transaction
+
+@Suppress("unused")
+class RegionsMutationService {
+
+    @GraphQLDescription("Updates the data privacy policy of a region")
+    fun updateDataPrivacy(regionId: Int, dataPrivacyText: String): Boolean {
+        transaction {
```
Comment (maxammann) [inline backend/src/main/kotlin/app/ehrenamtskarte/backend/regions/webservice/schema/RegionsMutationService.kt:12]: If a permission check missing here?

**Review by maxammann — APPROVED** [review]

**Comment by maxammann** [comment]
(Not tested yet)

File: administration/src/application/components/forms/ApplicationForm.tsx (hunk 3 above)
Comment (michael-markl) [reply administration/src/application/components/forms/ApplicationForm.tsx:12]: I think we do for blueprint. For mui, there might be some other options (and we might be inconsistent here)...
See https://mui.com/material-ui/guides/styled-engine/

File: administration/src/application/components/forms/ApplicationForm.tsx (hunk 4 above)
Comment (f1sh1918) [reply administration/src/application/components/forms/ApplicationForm.tsx:13]: for styled components yes, but i could pass it inline with the `sx` of mui.

File: administration/src/application/components/forms/ApplicationForm.tsx [hunk 7]
```diff
@@ -31,15 +40,21 @@ export type ApplicationFormState = {
   entitlementType: BlueCardEntitlementType | null
   standardEntitlement: StandardEntitlementFormState
   personalData: PersonalDataFormState
+  hasAcceptedPrivacyPolicy: boolean
+  openPrivacyPolicy: boolean
```
[the remark this answers is not in the document]
Comment (f1sh1918) [reply administration/src/application/components/forms/ApplicationForm.tsx]: moved to useState hook

File: administration/src/components/home/HomeController.tsx [hunk 8]
```diff
@@ -3,29 +3,45 @@ import { Button, H3 } from '@blueprintjs/core'
...
-        <NavLink to={'/users'}>
-          <Button style={{ margin: '10px' }} icon='people' text='Benutzer verwalten' />
-        </NavLink>
+        <>
+          <NavLink to={'/users'}>
+            <StyledButton icon='people' text='Benutzer verwalten' />
+          </NavLink>
+          <NavLink to={'/region'} reloadDocument>
+            <StyledButton icon='path-search' text='Region verwalten' />
+          </NavLink>
```
[the remark this answers is not in the document]
Comment (f1sh1918) [reply administration/src/components/home/HomeController.tsx:41]: ok i didn't know i can change the accessibility (also for the route) to region admins

File: administration/src/application/components/forms/ApplicationForm.tsx [hunk 9]
```diff
@@ -91,6 +106,33 @@ const ApplicationForm: Form<ApplicationFormState, Options, ValidatedInput, Addit
         state={state.personalData}
         setState={useUpdateStateCallback(setState, 'personalData')}
       />
+      <FormGroup row>
+        <Checkbox
+          checked={state.hasAcceptedPrivacyPolicy}
+          onChange={e => setState(state => ({ ...state, hasAcceptedPrivacyPolicy: e.target.checked }))}
+          required
+        />
```
[the remark this answers is not in the document]
Comment (f1sh1918) [reply administration/src/application/components/forms/ApplicationForm.tsx]: merged branch and resolved conflicts

File: backend/src/main/kotlin/app/ehrenamtskarte/backend/regions/database/repos/RegionsRepository.kt (hunk 5 above)
Comment (f1sh1918) [reply backend/src/main/kotlin/app/ehrenamtskarte/backend/regions/database/repos/RegionsRepository.kt:33]: Is it fine to restrict in the frontend textarea? Could add a char counter below

File: administration/src/application/components/ApplyController.tsx [hunk 10]
```diff
@@ -26,6 +27,10 @@ const ApplyController = () => {
     applicationStorageKey,
     lastCommitForApplicationForm
   )
+  const { loading: loadingPolicy, data: policyData } = useGetDataPolicyQuery({
```
Comment (michael-markl) [inline administration/src/application/components/ApplyController.tsx:30]: We need some error handling (ie. refetch button / error toast)... Maybe you can add a todo here. It's okay if we defer this until we have implemented the regionId query.

File: administration/src/application/components/ApplyController.tsx [hunk 11]
```diff
@@ -63,8 +68,16 @@ const ApplyController = () => {
     <div style={{ display: 'flex', justifyContent: 'center', alignItems: 'start', margin: '16px' }}>
       <div style={{ maxWidth: '1000px', width: '100%' }}>
         <h2 style={{ textAlign: 'center' }}>Blaue Ehrenamtskarte beantragen</h2>
-        <ApplicationForm.Component state={state} setState={setState} onSubmit={submit} loading={loading} />
-        <DialogActions>{loading ? null : <DiscardAllInputsButton discardAll={discardAll} />}</DialogActions>
+        <ApplicationForm.Component
+          state={state}
+          setState={setState}
+          onSubmit={submit}
+          loading={loading}
```
Comment (michael-markl) [inline administration/src/application/components/ApplyController.tsx:75]:
```suggestion
          loading={loading || loadingPolicy}
```

File: administration/src/application/components/forms/StepSendForm.tsx [hunk 12]
```diff
@@ -46,23 +50,45 @@ const StepSendForm: Form<StepSendFormState, Options, ValidatedInput, AdditionalP
...
+        <CheckboxForm.Component
+          label='Ich versichere, dass alle angegebenen Informationen korrekt und vollständig sind.'
+          state={state.givenInformationIsCorrectAndComplete}
+          setState={useUpdateStateCallback(setState, 'givenInformationIsCorrectAndComplete')}
+          options={givenInformationIsCorrectAndCompleteOptions}
+        />
+        <BasicDialog
+          open={openPrivacyPolicy}
+          maxWidth='lg'
+          onUpdateOpen={() => setOpenPrivacyPolicy(false)}
```
Comment (michael-markl) [inline administration/src/application/components/forms/StepSendForm.tsx:85]: Either make this `onClose` or use 
```suggestion
          onUpdateOpen={setOpenPrivacyPolicy}
```

File: administration/src/application/components/primitive-inputs/CheckboxForm.tsx [hunk 13]
```diff
@@ -22,7 +22,7 @@ const CheckboxForm: Form<CheckboxFormState, Options, ValidatedInput, AdditionalP
     const isInvalid = validationResult.type === 'error'
 
     return (
-      <FormGroup onBlur={() => setTouched(true)}>
```
Comment (michael-markl) [inline administration/src/application/components/primitive-inputs/CheckboxForm.tsx:25]: Why did you remove this?

File: administration/src/components/regions/RegionOverview.tsx [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+  display: flex;
+  flex-direction: row;
+  justify-content: flex-end;
+
+  & button {
+    margin: 5px;
+  }
+`
+
+const CharCounter = styled.span<{ hasError: boolean }>`
```
Comment (michael-markl) [inline administration/src/components/regions/RegionOverview.tsx:36]: Use `$hasError` so that styled components does not forward the prop to the DOM element

```suggestion
const CharCounter = styled.span<{ $hasError: boolean }>`
```

File: backend/src/main/kotlin/app/ehrenamtskarte/backend/regions/database/repos/RegionsRepository.kt (hunk 5 above)
Comment (michael-markl) [reply backend/src/main/kotlin/app/ehrenamtskarte/backend/regions/database/repos/RegionsRepository.kt:33]: I think we might also want to check this constraint (and conditionally throw) here, so that we catch this error instead of relying on the DB / Exposed to throw an error.

File: backend/src/main/kotlin/app/ehrenamtskarte/backend/regions/webservice/schema/RegionsMutationService.kt (hunk 6 above)
Comment (michael-markl) [reply backend/src/main/kotlin/app/ehrenamtskarte/backend/regions/webservice/schema/RegionsMutationService.kt:18]: LGTM

File: administration/src/components/home/HomeController.tsx [hunk 15]
```diff
@@ -3,29 +3,47 @@ import { Button, H3 } from '@blueprintjs/core'
...
+        <>
+          <NavLink to={'/users'}>
+            <StyledButton icon='people' text='Benutzer verwalten' />
+          </NavLink>
+        </>
+      ) : null}
+      {role === Role.RegionAdmin ? (
+        <NavLink to={'/region'} reloadDocument>
+          <StyledButton icon='path-search' text='Region verwalten' />
         </NavLink>
```
Comment (michael-markl) [inline administration/src/components/home/HomeController.tsx:44]: I think, if we add it here, we should probably also add it to the navigation bar. Otherwise it's pretty inconsistent.

File: administration/src/components/home/HomeController.tsx [hunk 16]
```diff
@@ -3,29 +3,47 @@ import { Button, H3 } from '@blueprintjs/core'
...
-        <NavLink to={'/users'}>
-          <Button style={{ margin: '10px' }} icon='people' text='Benutzer verwalten' />
+        <>
+          <NavLink to={'/users'}>
+            <StyledButton icon='people' text='Benutzer verwalten' />
+          </NavLink>
+        </>
+      ) : null}
+      {role === Role.RegionAdmin ? (
+        <NavLink to={'/region'} reloadDocument>
```
Comment (michael-markl) [inline administration/src/components/home/HomeController.tsx:42]: We shouldn't need `reloadDocument` here

File: administration/src/components/regions/RegionOverview.tsx [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+const Content = styled.div`
+  padding: 0 6rem;
+  width: 100%;
+  z-index: 0;
+  flex-grow: 1;
+  display: flex;
+  flex-wrap: wrap;
+  justify-content: space-evenly;
+  align-items: center;
+`
```
Comment (michael-markl) [inline administration/src/components/regions/RegionOverview.tsx:16]: This looks quite odd on "larger" screen resolutions (like my laptop has).
I think we need some `flex-direction: column` in `Content` to force the correct layout

File: administration/src/application/components/forms/StepSendForm.tsx [hunk 18]
```diff
@@ -46,23 +50,45 @@ const StepSendForm: Form<StepSendFormState, Options, ValidatedInput, AdditionalP
...
+    const PrivacyLabel = (
+      <div style={{ alignSelf: 'center' }}>
+        Ich erkläre mich damit einverstanden, dass meine Daten zum Zwecke der Antragsverarbeitung gespeichert werden und
+        akzeptiere die{' '}
+        <Button
+          variant='text'
+          style={{ textTransform: 'capitalize', padding: 0, verticalAlign: 'unset' }}
+          onClick={() => setOpenPrivacyPolicy(true)}>
+          Datenschutzerklärung
+        </Button>
```
Comment (michael-markl) [inline administration/src/application/components/forms/StepSendForm.tsx:64]:
```suggestion
        </Button>.
```

**Review by michael-markl — CHANGES_REQUESTED** [review]

File: administration/src/application/components/forms/ApplicationForm.tsx (hunk 3 above)
Comment (michael-markl) [reply administration/src/application/components/forms/ApplicationForm.tsx]: No, we don't, I think we can ignore this for now

File: administration/src/application/components/primitive-inputs/CheckboxForm.tsx (hunk 13 above)
Comment (f1sh1918) [reply administration/src/application/components/primitive-inputs/CheckboxForm.tsx:25]: because it executed the blur even i click on the Anchor of the data privacy. Dunno why we need this in that case

File: administration/src/components/home/HomeController.tsx (hunk 16 above)
Comment (f1sh1918) [reply administration/src/components/home/HomeController.tsx:42]: well i tried it and it cashes the api call without it

File: backend/src/main/kotlin/app/ehrenamtskarte/backend/regions/database/repos/RegionsRepository.kt (hunk 5 above)
Comment (maxammann) [reply backend/src/main/kotlin/app/ehrenamtskarte/backend/regions/database/repos/RegionsRepository.kt:33]: Yes, we should have a check in the graphql schema, else attackers are able to send megabytes of text to the database instead of only 20k chars.

**Review by maxammann — APPROVED** [review]
Looks good to me except for the missing length check of the privacy policy text.

File: backend/src/main/kotlin/app/ehrenamtskarte/backend/regions/database/repos/RegionsRepository.kt (hunk 5 above)
Comment (maxammann) [reply backend/src/main/kotlin/app/ehrenamtskarte/backend/regions/database/repos/RegionsRepository.kt:33]: You can just throw a RuntimeException if the text is too long.

Comment (f1sh1918) [reply backend/src/main/kotlin/app/ehrenamtskarte/backend/regions/database/repos/RegionsRepository.kt:33]: added a max char check according to the char length in the database and throw a exception!

File: administration/src/application/components/ApplyController.tsx (hunk 10 above)
Comment (f1sh1918) [reply administration/src/application/components/ApplyController.tsx:30]: added TODO

File: administration/src/components/home/HomeController.tsx (hunk 15 above)
Comment (michael-markl) [reply administration/src/components/home/HomeController.tsx]: Hmm yeah, it's very crowded. Maybe we'll merge user management into the Region Management tab later on. Let's keep it like this then for now.

File: administration/src/application/components/primitive-inputs/CheckboxForm.tsx (hunk 13 above)
Comment (michael-markl) [reply administration/src/application/components/primitive-inputs/CheckboxForm.tsx:25]: For a general Checkbox, the user should be hinted that it is required to check the checkbox once he moves his focus away from it without checking it. The case with the Datenschutzerklärungs-Dialog is quite special and I think we can live with the hint being shown even if the user only clicked on that button.
So I would propose to revert this small change.

File: administration/src/components/regions/RegionOverview.tsx (hunk 17 above)
Comment (f1sh1918) [reply administration/src/components/regions/RegionOverview.tsx]: i adjusted the style and added some margin and bigger headline to improve the layout

File: administration/src/application/components/ApplyController.tsx [hunk 19]
```diff
@@ -30,7 +30,22 @@ const SuccessContent = styled.div`
...
+    onError: error => {
+      console.error(error)
+      enqueueSnackbar('Beim Absenden des Antrags is ein Fehler aufgetreten', { variant: 'error' })
+    },
+    onCompleted: result => {
+      if (result) {
+        setState(() => ApplicationForm.initialState)
+        setFormSubmitted(true)
+      } else {
+        enqueueSnackbar('Beim Absenden des Antrags is ein Fehler aufgetreten.', { variant: 'error' })
```
Comment (michael-markl) [inline administration/src/application/components/ApplyController.tsx:45]:
```suggestion
        enqueueSnackbar('Beim Absenden des Antrags ist ein Fehler aufgetreten.', { variant: 'error' })
```

**Review by michael-markl — APPROVED** [review]
Very nice now!

**Comment by michael-markl** [comment]
@f1sh1918 I've just triggered the CI pipeline... somehow auto-triggering CircleCI is broken
